Library/Homebrew/utils/pypi.rb

```diff
@@ -0,0 +1,93 @@
+
+    begin
+      json = JSON.parse out
+    rescue JSON::ParserError
+      return
+    end
+
+    sdist = json["urls"].find { |url| url["packagetype"] == "sdist" }
+    [json["info"]["name"], sdist["url"], sdist["digests"]["sha256"]]
+  end
+
+  def unable_to_resolve_deps_error(formula_name, odie_on_error)
+    return unless odie_on_error
+
+    odie "Unable to resolve some dependencies. Please update the resources for \"#{formula_name}\" manually."
+  end
+
+  def update_python_resources!(formula, print_only: false, silent: false, odie_on_error: true)
+    # PyPI package name isn't always the same as the formula name. Try to infer from the URL.
+    pypi_name = if formula.stable.url.start_with?("https://files.pythonhosted.org/packages/")
+      File.basename(formula.stable.url).match(/^(.+)-[a-z0-9.]+$/)[1]
+    else
+      formula.name
+    end
+
+    # Ensure there are no non-pypi resources
+    non_pypi_resources = formula.resources.reject do |resource|
+      resource.url.start_with? "https://files.pythonhosted.org/packages/"
+    end
+    return unable_to_resolve_deps_error formula.name, odie_on_error unless non_pypi_resources.empty?
+
+    # Get list of required packages
+    ohai "Retrieving PyPI dependencies for \"#{pypi_name}==#{formula.version}\"" unless print_only || silent
+    pipgrip_output = Utils.popen_read Formula["pipgrip"].bin/"pipgrip", "--json", "--no-cache-dir",
+                                      "#{pypi_name}==#{formula.version}"
+    return unable_to_resolve_deps_error formula.name, odie_on_error unless $CHILD_STATUS.success?
+
+    packages = JSON.parse(pipgrip_output).sort.to_h
+
+    # Remove extra packages that may be included in pipgrip output
+    exclude_list = %w[argparse pip setuptools wsgiref] + [pypi_name]
+    packages.delete_if do |p|
+      exclude_list.include? p
+    end
+
+    # Create resource blocks
+    new_resource_blocks = ""
+    packages.each do |package, version|
+      name, url, checksum = get_pypi_info package, version
+      return unable_to_resolve_deps_error formula.name, odie_on_error if name.nil?
+
+      # Indent and append each resource block
+      new_resource_blocks += "  " + <<~EOS.chomp.gsub("\n", "\n  ") + "\n\n"
+        resource "#{name}" do
+          url "#{url}"
+          sha256 "#{checksum}"
+        end
+      EOS
+    end
+
+    # Print resource lines only if --print-only is passed
+    puts new_resource_blocks.chomp if print_only
+    return if print_only
+
+    # Check whether resources already exist (excluding homebrew-virtualenv)
+    if formula.resources.empty? ||
+       (formula.resources.length == 1 && formula.resources.first.name == "homebrew-virtualenv")
+      # Place resources above install method
+      inreplace_regex = /  def install/
+      new_resource_blocks += "  def install"
+    else
+      # Replace existing resource blocks with new resource blocks
+      inreplace_regex = /  (resource .* do\s+url .*\s+sha256 .*\s+ end\s*)+/
```

Comment:
The current method for replacing the existing resources is to use this regex to match all existing resources.

This regex doesn't work for cases like https://github.com/Homebrew/homebrew-core/pull/58557 where there are comments in between groups of resources. Currently, it will replace only the first group of resources.

Possible solutions I can think of:
1. Don't use a regex for this
    - This is probably idea but may not be possible. Is there another way to add/remove lines?
1. Change the regex to include these comments as well
    - I'd like to avoid a super complicated regex if possible (the existing one already isn't great)
1. Use the regex to remove all resource blocks. Then, separately, insert the new resource blocks as if none existed previously
    - I think this will lead to extra comments being left in random spots so it's not ideal
1. Fail if there is more than one group of resources
    - This is probably the best option (assuming the first isn't doable). This way, it forces the user to look at whatever is causing it to fail and resolve it manually. They can use `update-python-resources --print` to get the correct resource blocks and can then format it as needed.
    - This should probably also fail (or maybe just give a warning) if there's a comment before the first resource. Ex:
        ```ruby
        # This is special for some reason
        resource "special" do
          url "..."
          sha256 "..."
        end

        resource "normal" do ...
          url "..."
          sha256 "..."
        end
        ```

Thoughts? I think I'll go with 4 unless anyone has any ideas for 1